### PR TITLE
MVVM design pattern using zkoss framework.

### PR DESCRIPTION
**Pull request by jinishavora**
MVVM design pattern using zkoss framework.

This is for initial review only, Readme, Comments etc are pending.
@ohbus @iluwatar 
Just want to get some idea/suggestions/feedback for the code, if its acceptable then I can finish up the remaining pending tasks on this one.

### CODE REVIEW

**Review by ohbus — COMMENTED** [review]
Happy to see you increasing code coverage!

**Review by ohbus — CHANGES_REQUESTED** [review]
Looks good to me!

But there are a lot of xml configurations. Can you remove them if they are not necessary for achieving the context.

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:13]:
```suggestion
Model–View–Binder
```

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:17]: You can link to https://java-design-patterns.com/principles/#separation-of-concerns

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:27]: It might be too much to have both Android and zkoss implementation here. I would suggest picking only one of them.

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:201]:
```suggestion
* .NET framework applications
```

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:200]:
```suggestion
* Android apps
```

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:202]:
```suggestion
* JavaScript applications
```

File: model-view-viewmodel/README.md
Comment (iluwatar) [inline model-view-viewmodel/README.md:227]: Maybe these instructions should be placed next to the programmatic example?

File: model-view-viewmodel/pom.xml
Comment (iluwatar) [inline model-view-viewmodel/pom.xml:14]: In general, all the version numbers and at least the dependencies that are used elsewhere too should be placed in the root `pom.xml`.

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:14]: We should use the same versions of dependencies in all the patterns.

**Review by iluwatar — CHANGES_REQUESTED** [review]
@jinishavora thanks for taking the time to draft this pull request! I think the framework choice you've made is fine and it makes a good MVVM example. I added some comments to the pull request. It would be great if you could check them.

File: model-view-viewmodel/README.md
Comment (jinishavora) [reply model-view-viewmodel/README.md]: done

File: model-view-viewmodel/README.md
Comment (jinishavora) [reply model-view-viewmodel/README.md]: done

File: model-view-viewmodel/README.md
Comment (jinishavora) [reply model-view-viewmodel/README.md]: done

File: model-view-viewmodel/README.md
Comment (jinishavora) [reply model-view-viewmodel/README.md]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml:14]: done

Comment (ohbus) [inline model-view-viewmodel/pom.xml:14]:
```suggestion
		<commons-io.version>2.6</commons-io.version>
```

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:72]:
```suggestion
			<version>${commons-io.version}</version>
```

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:79]: Lombok is already available from the project root.

```suggestion
```

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:90]: I can see you have already added these dependencies in the project root already. 

And junit dependency is already present as well.

I would suggest you to remove them.

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:99]: Set version numbers using property tags

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:112]: Set version numbers using property tags. It is already there in the root pom

```suggestion
				<version>${compiler.version}</version>
```

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:122]: Set version using property

```suggestion
				<version>${maven-war-plugin.version}</version>
```

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:127]: It is already defined in the root pom [**here**](https://github.com/iluwatar/java-design-patterns/blob/b5ed8b227857f0b446157d43fa98814630922444/pom.xml#L427)

You need to correct the versioning

File: pom.xml
Comment (ohbus) [inline pom.xml:79]: Correct indentation

```suggestion
	    <commons-io.version>2.6</commons-io.version>
```

File: pom.xml
Comment (ohbus) [inline pom.xml:404]: Correct indentation and mention environment variables as versions.

```suggestion
            <dependency>
                <groupId>org.zkoss.zk</groupId>
                <artifactId>zkbind</artifactId>
                <version>${zk.version}</version>
            </dependency>
            <dependency>
                <groupId>org.zkoss.zk</groupId>
                <artifactId>zul</artifactId>
                <version>${zk.version}</version>
            </dependency>
            <dependency>
                <groupId>org.zkoss.zk</groupId>
                <artifactId>zkplus</artifactId>
                <version>${zk.version}</version>
            </dependency>
            <dependency>
                <groupId>org.zkoss.zk</groupId>
                <artifactId>zhtml</artifactId>
                <version>${zk.version}</version>
            </dependency>
            <dependency>
                <groupId>commons-io</groupId>
                <artifactId>commons-io</artifactId>
                <version>${commons-io.version}</version>
            </dependency>
```

**Review by ohbus — CHANGES_REQUESTED** [review]
As you are in a pursuit of making a perfect Pull Request I would suggest you to ponder on the few remaining formatting changes that I have suggested in the comments.

Please request a review from @iluwatar and me once you have pushed the changes and are ready for mering.

Looking forward to it. Awesome work 🥂

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml:90]: @ohbus If I remove these dependencies from the module's pom, the module as a separate individual project won't compile. 
Are we ok with that?

Comment (ohbus) [reply model-view-viewmodel/pom.xml:90]: You can either put it here if necessary or in the parent pom.

Do not put same dependencies in parent and module pom.

And you can safely remove dependencies that are actually present in the parent pom like lombok.

Comment (jinishavora) [reply model-view-viewmodel/pom.xml:90]: Have made changes in parent pom and module pom, such that module can be compiled individually and whatever can be part of parent pom is in parent and what is required for module to compile successfully is in it.

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: pom.xml
Comment (jinishavora) [reply pom.xml]: done

File: pom.xml
Comment (jinishavora) [reply pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: couldn't find the version in root pom.xml, so updated in module pom.xml accordingly.

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml]: done

**Comment by jinishavora** [comment]
@ohbus @iluwatar All suggestions implemented/updated accordingly, please review.

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:62]: As I already mentioned the dependency exists in the root [**pom**](https://github.com/iluwatar/java-design-patterns/blob/b5ed8b227857f0b446157d43fa98814630922444/pom.xml#L298).

```suggestion
```

https://github.com/iluwatar/java-design-patterns/blob/b5ed8b227857f0b446157d43fa98814630922444/pom.xml#L298

File: model-view-viewmodel/pom.xml
Comment (ohbus) [inline model-view-viewmodel/pom.xml:16]: The guava version already exists in the root pom over [here](url)

I will suggest you to remove it.

https://github.com/iluwatar/java-design-patterns/blob/b5ed8b227857f0b446157d43fa98814630922444/pom.xml#L53

File: pom.xml
Comment (ohbus) [inline pom.xml:78]: Correct the indentation here please.

```suggestion
	<commons-io.version>2.6</commons-io.version>
```

**Review by ohbus — CHANGES_REQUESTED** [review]
Thank you for the updates, but some of them have been overlooked.

Please take a look at the comments below.

File: model-view-viewmodel/pom.xml
Comment (iluwatar) [inline model-view-viewmodel/pom.xml:32]: Having a separate LGPL licensed module in the project might be problematic. I think we need to look into this and discuss it a bit more.

**Review by iluwatar — CHANGES_REQUESTED** [review]

File: model-view-viewmodel/pom.xml
Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: See this link: https://www.zkoss.org/license

Comment (ohbus) [reply model-view-viewmodel/pom.xml:32]: Is the pattern licensed using GPL3 or is it the library zkoss is?

@jinishavora

Comment (ohbus) [reply model-view-viewmodel/pom.xml:32]: > See this link: https://www.zkoss.org/license

So we can use it over here, but as per GPL I believe we need to leave it as it is.

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: I believe having this new LGPL-licensed module as a dependency would make our project what they 'combined work' in https://www.gnu.org/licenses/lgpl-3.0.html and we would have to fulfill those requirements.

Comment (ohbus) [reply model-view-viewmodel/pom.xml:32]: Well we are not modifying any part of the framework, but using it. So I guess we are okay including it in the pom and be done with it right?

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: If the sample application using the framework really is LGPL licensed, I think it's a bit more complicated than that. We would at least have to

- [ ] Give prominent notice with each copy of the Combined Work that the Library is used in it and that the Library and its use are covered by this License.
- [ ] Accompany the Combined Work with a copy of the GNU GPL and this license document.
- [ ] Configure license-maven-plugin so that it puts a different license to the files under model-view-viewmodel

Comment (ohbus) [reply model-view-viewmodel/pom.xml:32]: So this module need to follow the LGPL license while the rest of the project will continue with MIT?

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: Yes, that's at least how I understand it. I think the above criteria should be fulfilled by this pull request. An alternative approach would be getting rid of zkoss framework and creating the pattern without it or with something else.

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: One consideration is also how this affects our "end-users". Currently, everything is published under the permissive MIT license. I'm not strict with it, but if one reads and interprets it very carefully and literally, copying and pasting code from this repository to some other project would require the copier to MIT license the copied portion (see for example the discussion [here](https://github.com/salesforce/gorp/issues/2)). If we include this LGPL licensed module, it becomes more confusing from the end user's point of view due to different licensing rules depending on where you take the code from.

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml:62]: @ohbus, this has not been overlooked, I am well aware of its presence.
If I remove it, the individual module doesn't compile.
I can't remove it from parent either as its being used by other modules as well.

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml:16]: @ohbus, As mentioned in above comment same is the case with guava.
If I remove it, the individual module doesn't compile.
I can't remove it from parent either as its being used by other modules as well.

File: model-view-viewmodel/pom.xml
Comment (jinishavora) [reply model-view-viewmodel/pom.xml:32]: What's the verdict here? Do we want to keep this as is or not use zkoss and find something else with MIT license.
I am not really sure if we can find something for this pattern under MIT license.

Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: I'm thinking that we can merge the pattern as is and fulfill the LGPL licensing requirements in a follow-up ticket. It's a permissive license anyway, we just need to make some adjustments here and there.

Comment (jinishavora) [reply model-view-viewmodel/pom.xml:32]: Sounds good!
Once approved, I can merge this and then make the changes required in the follow-up.

**Review by iluwatar — APPROVED** [review]
On my behalf, this can be merged once all the current review comments have been addressed.

**Comment by jinishavora** [comment]
Thanks @iluwatar!
I am waiting on @ohbus to check my comments on his suggested changes and revert.
Once we decide on changes/no changes and he approves, I shall merge this.

File: model-view-viewmodel/pom.xml
Comment (ohbus) [reply model-view-viewmodel/pom.xml:62]: Okay, it should not have been the case though.

File: model-view-viewmodel/pom.xml
Comment (ohbus) [reply model-view-viewmodel/pom.xml:16]: I believe the code should be able to meet the dependency requirements from the parent pom just like all other modules in this repository does.

I will follow it up after we merge the code.

File: model-view-viewmodel/pom.xml
Comment (ohbus) [reply model-view-viewmodel/pom.xml:32]: Sounds like a good idea.

**Review by ohbus — APPROVED** [review]
We need to follow up with the LGPLv3 license for getting compliant with the ZKoss framework.

And will also check with the dependeis like JUnit and Guava to use it from the parent pom.

**Comment by ohbus** [comment]
Also resolves #601

**Comment by ohbus** [comment]
Thank you @iluwatar for helping with the review.

@all-contributors please add @jinishavora for code

File: model-view-viewmodel/pom.xml
Comment (iluwatar) [reply model-view-viewmodel/pom.xml:32]: Follow-up licensing task: https://github.com/iluwatar/java-design-patterns/issues/1690

**Pull request merged**
